contracts/validator/contract/XDCValidatorV2.sol

```diff
@@ -0,0 +1,531 @@
+        payable
+        onlyValidCandidateCap
+        onlyKYCWhitelisted
+        onlyNotCandidate(_candidate)
+    {
+        uint256 cap = validatorsState[_candidate].cap.add(msg.value);
+        candidates.push(_candidate);
+        validatorsState[_candidate] = ValidatorState({
+            owner: msg.sender,
+            isCandidate: true,
+            cap: cap
+        });
+        validatorsState[_candidate].voters[msg.sender] = validatorsState[
+            _candidate
+        ].voters[msg.sender].add(msg.value);
+        candidateCount = candidateCount.add(1);
+        if (ownerToCandidate[msg.sender].length == 0) {
+            owners.push(msg.sender);
+            ownerCount++;
+        }
+        ownerToCandidate[msg.sender].push(_candidate);
+        voters[_candidate].push(msg.sender);
+        emit Propose(msg.sender, _candidate, msg.value);
+    }
+
+    function vote(
+        address _candidate
+    ) external payable onlyValidVoterCap onlyValidCandidate(_candidate) {
+        validatorsState[_candidate].cap = validatorsState[_candidate].cap.add(
+            msg.value
+        );
+        if (validatorsState[_candidate].voters[msg.sender] == 0) {
+            voters[_candidate].push(msg.sender);
+        }
+        validatorsState[_candidate].voters[msg.sender] = validatorsState[
+            _candidate
+        ].voters[msg.sender].add(msg.value);
+        emit Vote(msg.sender, _candidate, msg.value);
+    }
+
+    function getCandidates() public view returns (address[]) {
+        return candidates;
+    }
+
+    function getCandidateCap(address _candidate) public view returns (uint256) {
+        return validatorsState[_candidate].cap;
+    }
+
+    function getCandidateOwner(
+        address _candidate
+    ) public view returns (address) {
+        return validatorsState[_candidate].owner;
+    }
+
+    function getVoterCap(
+        address _candidate,
+        address _voter
+    ) public view returns (uint256) {
+        return validatorsState[_candidate].voters[_voter];
+    }
+
+    function getVoters(address _candidate) public view returns (address[]) {
+        return voters[_candidate];
+    }
+
+    function isCandidate(address _candidate) public view returns (bool) {
+        return validatorsState[_candidate].isCandidate;
+    }
+
+    function getWithdrawBlockNumbers() public view returns (uint256[]) {
+        return withdrawsState[msg.sender].blockNumbers;
+    }
+
+    function getWithdrawCap(
+        uint256 _blockNumber
+    ) public view returns (uint256) {
+        return withdrawsState[msg.sender].caps[_blockNumber];
+    }
+
+    function unvote(
+        address _candidate,
+        uint256 _cap
+    ) public onlyValidVote(_candidate, _cap) {
+        validatorsState[_candidate].cap = validatorsState[_candidate].cap.sub(
+            _cap
+        );
+        validatorsState[_candidate].voters[msg.sender] = validatorsState[
+            _candidate
+        ].voters[msg.sender].sub(_cap);
```

Comment:
if now the cap is 0, should we clean it?

Reply:
if cap is 0 , just sub 0